### ISSUE TITLE
Failed to synthesize programs using a self-trained model

### ISSUE THREAD

**Issue by ylu1317** [capergroup/bayou#215]
Hello,

I'm trying to use Bayou to synthesize a program that uses [Jsoup](https://jsoup.org) like this:
```
Document doc = Jsoup.parse(content);
Elements links = doc.select(selector);
Element link = links.first();
String href = link.attr(attr);
```

I downloaded the Jsoup source code and some API usages of Jsoup, followed the steps [here](https://github.com/capergroup/bayou/issues/208) and finally finished training my own model using only source code I mentioned above. Then I followed the main page of Bayou and built Bayou successfully. Before that, I added the Jsoup JAR file into the class path of the DOM driver at [here](https://github.com/ylu1317/bayou/blob/fb25fdb7e53bdbc0f45b82fd6095eda2f22a70ac/src/main/java/edu/rice/cs/caper/bayou/application/dom_driver/Driver.java#L53) and also the Parser at [here](https://github.com/ylu1317/bayou/blob/fb25fdb7e53bdbc0f45b82fd6095eda2f22a70ac/src/main/java/edu/rice/cs/caper/bayou/core/synthesizer/Parser.java#L71) and [here](https://github.com/ylu1317/bayou/blob/fb25fdb7e53bdbc0f45b82fd6095eda2f22a70ac/src/main/java/edu/rice/cs/caper/bayou/core/synthesizer/Parser.java#L117) to make sure all the Jsoup JAR file is fed into Bayou.

Finally, I start the bayou server using
```
python3 $SCRIPT_DIR/python/bayou/server/ast_server.py --save_dir $MODEL_DIR --logs_dir $LOG_DIR
```
where `MODEL_DIR` is the model I trained using the JSoup source code and some API usages. The script shows that it was using the model I provided.

However, when I ran `synthesize.sh` and provided the following program:
```
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class HTML {
    public void read_text(String content, String selector, String attr) {
        {
            ///call:parse type:Document
        }
    }

}
```
I got this error message:
```
Exception in thread "main" java.io.IOException: Unexpected status code: 500
	at edu.rice.cs.caper.bayou.core.bayou_services_client.api_synthesis.ApiSynthesisClient.synthesizeHelp(ApiSynthesisClient.java:142)
	at edu.rice.cs.caper.bayou.core.bayou_services_client.api_synthesis.ApiSynthesisClient.synthesise(ApiSynthesisClient.java:80)
	at edu.rice.cs.caper.bayou.application.api_synthesis_server.ApiSynthesisLocalClient.synthesise(ApiSynthesisLocalClient.java:57)
	at edu.rice.cs.caper.bayou.application.api_synthesis_server.ApiSynthesisLocalClient.main(ApiSynthesisLocalClient.java:132)
```

I suspect that it was my own model, because when I switched using the model that comes with Bayou, everything works well (except the fact that I didn't get the results I want since the default model was trained using Jsoup). 

The error message seems to be caused by the synthesis server, but I cannot find any error message generated by the synthesis server so I really don't know what to do at this point. Do you guys mind pointing me to something that at least shows what the error is?

Thank you guys very much and let me know if you guys need anything else because I'm not sure if I included everything in this post.

**Comment by vijay-murali** [comment]
To reproduce the error I would need the model, of course. But before that, do you see any sketches being generated in `logs/ast_server.log`? If so the AST server is responding to queries correctly, and the problem is most likely in the synthesizer. Otherwise we will have to look into the AST server code.

**Comment by ylu1317** [comment]
Here is the first entry in the `logs/ast_server.log`:
```
20-07-2018:21:47:41,452 DEBUG    [MainThread ast_server.py:146] entering
20-07-2018:21:47:41,466 INFO     [MainThread ast_server.py:153] loading model
20-07-2018:21:47:46,396 INFO     [MainThread tf_logging.py:115] Restoring parameters from /root/html/model/model99.ckpt
20-07-2018:21:47:46,793 INFO     [MainThread _internal.py:88]  * Running on http://0.0.0.0:8084/ (Press CTRL+C to quit)
20-07-2018:22:17:18,180 DEBUG    [MainThread ast_server.py:146] entering
20-07-2018:22:17:18,193 INFO     [MainThread ast_server.py:153] loading model
20-07-2018:22:17:23,98 INFO     [MainThread tf_logging.py:115] Restoring parameters from /root/html/model/model99.ckpt
20-07-2018:22:17:23,275 INFO     [MainThread _internal.py:88]  * Running on http://0.0.0.0:8084/ (Press CTRL+C to quit)
20-07-2018:22:17:33,322 DEBUG    [Thread-1 ast_server.py:34] request_json:{
  "evidence": "{\n  \"apicalls\": [\n    \"add\"\n  ],\n  \"types\": [\n    \"List\",\n    \"String\"\n  ],\n  \"keywords\": []\n}",
  "request type": "generate asts"
}
20-07-2018:22:17:33,322 DEBUG    [Thread-1 ast_server.py:64] entering
20-07-2018:22:17:37,843 DEBUG    [Thread-1 ast_server.py:91] exiting
20-07-2018:22:17:37,844 DEBUG    [Thread-1 ast_server.py:59] {
  "evidences": {
    "apicalls": [
      "add"
    ],
    "types": [
      "List",
      "String"
    ],
    "keywords": [
      "list",
      "add",
      "string"
    ]
  },
  "asts": [
    {
      "ast": {
        "node": "DSubTree",
        "_nodes": [
          {
            "node": "DAPICall",
            "_call": "org.jsoup.Connection.Base<Tau_T>.url(java.net.URL)"
          }
        ]
      },
      "probability": "1.000000e+00"
    },
    {
      "ast": {
        "node": "DSubTree",
        "_nodes": [
          {
            "node": "DAPICall",
            "_call": "org.jsoup.Connection.Base<Tau_T>.url(java.net.URL)"
          },
          {
            "node": "DAPICall",
            "_call": "DSubTree"
          }
        ]
      },
      "probability": "0.000000e+00"
    }
  ]
}
20-07-2018:22:17:37,846 INFO     [Thread-1 _internal.py:88] 127.0.0.1 - - [20/Jul/2018 22:17:37] "POST / HTTP/1.1" 200 -
```
The rest of the entries look similar to this one, and I realize that the `probability` fields are all zero. Hope this helps. Thanks!

**Comment by vijay-murali** [comment]
Looks like sketches are being generated, so the AST server is responding well. However, the probabilities being 1 or 0 is suspicious, and suggests some problem with the model training. I would need more details about the training run and configuration. Can you post the model config file and the loss during training epochs (to get the loss grep for "Model checkpoint" in the training log file).

PS: If you want to talk we can take this offline and post the final solution here.

**Comment by ylu1317** [comment]
Here is the model config file:
```
{
    "model": "lle",
        "latent_size": 32,
        "batch_size": 50,
        "num_epochs": 100,
        "learning_rate": 0.02,
        "print_step": 1,
        "alpha": 1e-05,
        "beta": 1e-05,
        "evidence": [{
            "name": "apicalls",
            "units": 64,
            "num_layers": 3,
            "tile": 1
        },
        {
            "name": "types",
            "units": 32,
            "num_layers": 3,
            "tile": 1
        },
        {
            "name": "keywords",
            "units": 64,
            "num_layers": 3,
            "tile": 1
        }
        ],
        "decoder": {
            "units": 256,
            "num_layers": 3,
            "max_ast_depth": 32
        }
}
```
And here is the output of `grep "Model checkpoint"` on the log file:
```
Model checkpointed: checkpoint/model0.ckpt. Average for epoch evidence: -0.010, latent: 0.001, generation: 0.508, loss: 0.499
Model checkpointed: checkpoint/model1.ckpt. Average for epoch evidence: -0.016, latent: 0.001, generation: 0.189, loss: 0.174
Model checkpointed: checkpoint/model2.ckpt. Average for epoch evidence: -0.023, latent: 0.001, generation: 0.161, loss: 0.138
Model checkpointed: checkpoint/model3.ckpt. Average for epoch evidence: -0.017, latent: 0.001, generation: 0.144, loss: 0.127
Model checkpointed: checkpoint/model4.ckpt. Average for epoch evidence: -0.020, latent: 0.001, generation: 0.122, loss: 0.102
Model checkpointed: checkpoint/model5.ckpt. Average for epoch evidence: -0.019, latent: 0.001, generation: 0.112, loss: 0.094
Model checkpointed: checkpoint/model6.ckpt. Average for epoch evidence: -0.022, latent: 0.001, generation: 0.089, loss: 0.067
Model checkpointed: checkpoint/model7.ckpt. Average for epoch evidence: -0.023, latent: 0.001, generation: 0.075, loss: 0.052
Model checkpointed: checkpoint/model8.ckpt. Average for epoch evidence: -0.022, latent: 0.001, generation: 0.075, loss: 0.054
Model checkpointed: checkpoint/model9.ckpt. Average for epoch evidence: -0.018, latent: 0.001, generation: 0.077, loss: 0.060
Model checkpointed: checkpoint/model10.ckpt. Average for epoch evidence: -0.006, latent: 0.001, generation: 0.104, loss: 0.099
Model checkpointed: checkpoint/model11.ckpt. Average for epoch evidence: 0.008, latent: 0.001, generation: 0.311, loss: 0.320
Model checkpointed: checkpoint/model12.ckpt. Average for epoch evidence: 0.158, latent: 0.002, generation: 10.481, loss: 10.641
Model checkpointed: checkpoint/model13.ckpt. Average for epoch evidence: 0.205, latent: 0.008, generation: 45.957, loss: 46.170
Model checkpointed: checkpoint/model14.ckpt. Average for epoch evidence: 0.296, latent: 0.011, generation: 41.426, loss: 41.732
Model checkpointed: checkpoint/model15.ckpt. Average for epoch evidence: 0.328, latent: 0.031, generation: 85.310, loss: 85.669
Model checkpointed: checkpoint/model16.ckpt. Average for epoch evidence: 0.294, latent: 0.026, generation: 97.353, loss: 97.673
Model checkpointed: checkpoint/model17.ckpt. Average for epoch evidence: 0.212, latent: 0.014, generation: 51.672, loss: 51.899
Model checkpointed: checkpoint/model18.ckpt. Average for epoch evidence: 0.312, latent: 0.023, generation: 55.039, loss: 55.373
Model checkpointed: checkpoint/model19.ckpt. Average for epoch evidence: 0.253, latent: 0.019, generation: 35.813, loss: 36.085
Model checkpointed: checkpoint/model20.ckpt. Average for epoch evidence: 0.218, latent: 0.015, generation: 27.455, loss: 27.688
Model checkpointed: checkpoint/model21.ckpt. Average for epoch evidence: 0.205, latent: 0.013, generation: 24.364, loss: 24.582
Model checkpointed: checkpoint/model22.ckpt. Average for epoch evidence: 0.432, latent: 0.046, generation: 96.918, loss: 97.397
Model checkpointed: checkpoint/model23.ckpt. Average for epoch evidence: 0.647, latent: 0.081, generation: 152.303, loss: 153.031
Model checkpointed: checkpoint/model24.ckpt. Average for epoch evidence: 0.576, latent: 0.055, generation: 99.908, loss: 100.539
Model checkpointed: checkpoint/model25.ckpt. Average for epoch evidence: 0.437, latent: 0.027, generation: 62.038, loss: 62.501
Model checkpointed: checkpoint/model26.ckpt. Average for epoch evidence: 3.779, latent: 0.020, generation: 49.937, loss: 53.737
Model checkpointed: checkpoint/model27.ckpt. Average for epoch evidence: 3.129, latent: 0.014, generation: 52.264, loss: 55.407
Model checkpointed: checkpoint/model28.ckpt. Average for epoch evidence: 0.636, latent: 0.011, generation: 50.750, loss: 51.397
Model checkpointed: checkpoint/model29.ckpt. Average for epoch evidence: 0.413, latent: 0.010, generation: 42.568, loss: 42.991
Model checkpointed: checkpoint/model30.ckpt. Average for epoch evidence: 0.351, latent: 0.008, generation: 23.035, loss: 23.395
Model checkpointed: checkpoint/model31.ckpt. Average for epoch evidence: 0.342, latent: 0.007, generation: 13.699, loss: 14.049
Model checkpointed: checkpoint/model32.ckpt. Average for epoch evidence: 0.343, latent: 0.007, generation: 8.821, loss: 9.171
Model checkpointed: checkpoint/model33.ckpt. Average for epoch evidence: 0.340, latent: 0.007, generation: 6.778, loss: 7.125
Model checkpointed: checkpoint/model34.ckpt. Average for epoch evidence: 0.338, latent: 0.007, generation: 5.497, loss: 5.843
Model checkpointed: checkpoint/model35.ckpt. Average for epoch evidence: 0.333, latent: 0.007, generation: 4.810, loss: 5.150
Model checkpointed: checkpoint/model36.ckpt. Average for epoch evidence: 0.332, latent: 0.007, generation: 4.717, loss: 5.056
Model checkpointed: checkpoint/model37.ckpt. Average for epoch evidence: 0.341, latent: 0.007, generation: 3.885, loss: 4.233
Model checkpointed: checkpoint/model38.ckpt. Average for epoch evidence: 0.345, latent: 0.007, generation: 3.926, loss: 4.278
Model checkpointed: checkpoint/model39.ckpt. Average for epoch evidence: 0.344, latent: 0.007, generation: 3.661, loss: 4.011
Model checkpointed: checkpoint/model40.ckpt. Average for epoch evidence: 0.351, latent: 0.007, generation: 8.595, loss: 8.952
Model checkpointed: checkpoint/model41.ckpt. Average for epoch evidence: 0.187, latent: 0.015, generation: 47.826, loss: 48.029
Model checkpointed: checkpoint/model42.ckpt. Average for epoch evidence: 0.153, latent: 0.019, generation: 30.518, loss: 30.690
Model checkpointed: checkpoint/model43.ckpt. Average for epoch evidence: 0.137, latent: 0.015, generation: 23.488, loss: 23.640
Model checkpointed: checkpoint/model44.ckpt. Average for epoch evidence: 0.126, latent: 0.015, generation: 22.892, loss: 23.033
Model checkpointed: checkpoint/model45.ckpt. Average for epoch evidence: 0.122, latent: 0.016, generation: 22.374, loss: 22.511
Model checkpointed: checkpoint/model46.ckpt. Average for epoch evidence: 0.121, latent: 0.016, generation: 19.321, loss: 19.457
Model checkpointed: checkpoint/model47.ckpt. Average for epoch evidence: 0.120, latent: 0.015, generation: 16.416, loss: 16.551
Model checkpointed: checkpoint/model48.ckpt. Average for epoch evidence: 0.121, latent: 0.015, generation: 15.406, loss: 15.541
Model checkpointed: checkpoint/model49.ckpt. Average for epoch evidence: 0.119, latent: 0.014, generation: 15.508, loss: 15.641
Model checkpointed: checkpoint/model50.ckpt. Average for epoch evidence: 0.123, latent: 0.014, generation: 20.028, loss: 20.165
Model checkpointed: checkpoint/model51.ckpt. Average for epoch evidence: 0.160, latent: 0.019, generation: 43.486, loss: 43.665
Model checkpointed: checkpoint/model52.ckpt. Average for epoch evidence: 0.163, latent: 0.018, generation: 36.808, loss: 36.989
Model checkpointed: checkpoint/model53.ckpt. Average for epoch evidence: 0.191, latent: 0.017, generation: 36.075, loss: 36.284
Model checkpointed: checkpoint/model54.ckpt. Average for epoch evidence: 0.189, latent: 0.016, generation: 30.117, loss: 30.322
Model checkpointed: checkpoint/model55.ckpt. Average for epoch evidence: 0.180, latent: 0.016, generation: 22.086, loss: 22.281
Model checkpointed: checkpoint/model56.ckpt. Average for epoch evidence: 0.176, latent: 0.015, generation: 18.369, loss: 18.560
Model checkpointed: checkpoint/model57.ckpt. Average for epoch evidence: 0.169, latent: 0.015, generation: 14.182, loss: 14.365
Model checkpointed: checkpoint/model58.ckpt. Average for epoch evidence: 0.164, latent: 0.015, generation: 13.233, loss: 13.411
Model checkpointed: checkpoint/model59.ckpt. Average for epoch evidence: 0.174, latent: 0.018, generation: 29.786, loss: 29.978
Model checkpointed: checkpoint/model60.ckpt. Average for epoch evidence: 0.237, latent: 0.026, generation: 65.730, loss: 65.993
Model checkpointed: checkpoint/model61.ckpt. Average for epoch evidence: 0.191, latent: 0.022, generation: 67.178, loss: 67.390
Model checkpointed: checkpoint/model62.ckpt. Average for epoch evidence: 0.150, latent: 0.018, generation: 36.708, loss: 36.876
Model checkpointed: checkpoint/model63.ckpt. Average for epoch evidence: 0.132, latent: 0.017, generation: 21.301, loss: 21.450
Model checkpointed: checkpoint/model64.ckpt. Average for epoch evidence: 0.125, latent: 0.016, generation: 14.560, loss: 14.701
Model checkpointed: checkpoint/model65.ckpt. Average for epoch evidence: 0.125, latent: 0.015, generation: 11.199, loss: 11.338
Model checkpointed: checkpoint/model66.ckpt. Average for epoch evidence: 0.124, latent: 0.014, generation: 8.753, loss: 8.891
Model checkpointed: checkpoint/model67.ckpt. Average for epoch evidence: 0.122, latent: 0.014, generation: 7.492, loss: 7.628
Model checkpointed: checkpoint/model68.ckpt. Average for epoch evidence: 0.120, latent: 0.013, generation: 7.142, loss: 7.276
Model checkpointed: checkpoint/model69.ckpt. Average for epoch evidence: 0.119, latent: 0.013, generation: 6.108, loss: 6.240
Model checkpointed: checkpoint/model70.ckpt. Average for epoch evidence: 0.117, latent: 0.013, generation: 6.171, loss: 6.301
Model checkpointed: checkpoint/model71.ckpt. Average for epoch evidence: 0.115, latent: 0.013, generation: 5.885, loss: 6.012
Model checkpointed: checkpoint/model72.ckpt. Average for epoch evidence: 0.107, latent: 0.013, generation: 6.123, loss: 6.243
Model checkpointed: checkpoint/model73.ckpt. Average for epoch evidence: 0.102, latent: 0.013, generation: 6.589, loss: 6.703
Model checkpointed: checkpoint/model74.ckpt. Average for epoch evidence: 0.101, latent: 0.012, generation: 5.302, loss: 5.415
Model checkpointed: checkpoint/model75.ckpt. Average for epoch evidence: 0.100, latent: 0.012, generation: 4.521, loss: 4.633
Model checkpointed: checkpoint/model76.ckpt. Average for epoch evidence: 0.099, latent: 0.012, generation: 4.011, loss: 4.122
Model checkpointed: checkpoint/model77.ckpt. Average for epoch evidence: 0.099, latent: 0.012, generation: 6.846, loss: 6.956
Model checkpointed: checkpoint/model78.ckpt. Average for epoch evidence: 0.098, latent: 0.011, generation: 4.936, loss: 5.045
Model checkpointed: checkpoint/model79.ckpt. Average for epoch evidence: 0.098, latent: 0.011, generation: 3.763, loss: 3.872
Model checkpointed: checkpoint/model80.ckpt. Average for epoch evidence: 0.098, latent: 0.011, generation: 3.357, loss: 3.466
Model checkpointed: checkpoint/model81.ckpt. Average for epoch evidence: 0.098, latent: 0.011, generation: 3.044, loss: 3.152
Model checkpointed: checkpoint/model82.ckpt. Average for epoch evidence: 0.096, latent: 0.011, generation: 3.143, loss: 3.250
Model checkpointed: checkpoint/model83.ckpt. Average for epoch evidence: 0.094, latent: 0.011, generation: 3.235, loss: 3.340
Model checkpointed: checkpoint/model84.ckpt. Average for epoch evidence: 0.095, latent: 0.010, generation: 3.356, loss: 3.462
Model checkpointed: checkpoint/model85.ckpt. Average for epoch evidence: 0.100, latent: 0.011, generation: 4.932, loss: 5.043
Model checkpointed: checkpoint/model86.ckpt. Average for epoch evidence: 0.106, latent: 0.011, generation: 5.077, loss: 5.194
Model checkpointed: checkpoint/model87.ckpt. Average for epoch evidence: 0.997, latent: 0.014, generation: 41.206, loss: 42.216
Model checkpointed: checkpoint/model88.ckpt. Average for epoch evidence: 0.989, latent: 0.038, generation: 356.586, loss: 357.613
Model checkpointed: checkpoint/model89.ckpt. Average for epoch evidence: 0.258, latent: 0.027, generation: 296.136, loss: 296.421
Model checkpointed: checkpoint/model90.ckpt. Average for epoch evidence: 0.274, latent: 0.016, generation: 182.244, loss: 182.535
Model checkpointed: checkpoint/model91.ckpt. Average for epoch evidence: 0.296, latent: 0.030, generation: 208.532, loss: 208.858
Model checkpointed: checkpoint/model92.ckpt. Average for epoch evidence: 0.380, latent: 0.093, generation: 717.854, loss: 718.327
Model checkpointed: checkpoint/model93.ckpt. Average for epoch evidence: 0.338, latent: 0.077, generation: 842.094, loss: 842.510
Model checkpointed: checkpoint/model94.ckpt. Average for epoch evidence: 0.303, latent: 0.058, generation: 562.999, loss: 563.360
Model checkpointed: checkpoint/model95.ckpt. Average for epoch evidence: 0.286, latent: 0.045, generation: 366.232, loss: 366.564
Model checkpointed: checkpoint/model96.ckpt. Average for epoch evidence: 0.277, latent: 0.038, generation: 249.317, loss: 249.632
Model checkpointed: checkpoint/model97.ckpt. Average for epoch evidence: 0.271, latent: 0.033, generation: 188.573, loss: 188.876
Model checkpointed: checkpoint/model98.ckpt. Average for epoch evidence: 0.266, latent: 0.029, generation: 141.710, loss: 142.005
Model checkpointed: checkpoint/model99.ckpt. Average for epoch evidence: 0.263, latent: 0.026, generation: 113.827, loss: 114.116
```

**Comment by vijay-murali** [comment]
The problem is with the model training. The loss should converge to a small value (typically 0.3 or less) in the end but it keeps diverging after 10 epochs. This usually indicates the model is too complex for the data. I would suggest using the config file for the current Bayou model to train yours too, adjusting the number of epochs based on your data size.

**Comment by ylu1317** [comment]
After I discussed with Vijay, I changed the all the `num_layers` fields in the config file to 1 and set the `num_epochs` to 50 and everything worked smoothly. The problem was that the model was overtrained which was showed in the loss log file I posted. As Vijay suggested, the number of layers should typically set to 1 and the training should stop when loss is around 0.05 - 0.1 (correct me if I'm wrong Vijay). Normally, the training should converge around 30-50 epochs. Another thing Vijay suggested is to use the training config file used by the default model.
